### PR TITLE
OS Managed Power Management framework

### LINKED ISSUE #8255: [RFC] Add support for system suspend/resume handling from kernel
Add the necessary PM hooks and framework in the kernel to handle system suspend and resume flows.

The goal is to move the common PM logic to kernel make the application less bother about the internal details.

### PR DESCRIPTION
**Pull request by ramakrishnapallala**
arch: arm: nordic_nrf: Add an API to check for valid PM state
subsys: power: Add support for OS managed Power Management framework
samples: subsys: Add sample app to demo OS managed PM

Fixes #8255

### CODE REVIEW

**Review by pizi-nordic — DISMISSED** [review]

File: samples/boards/nrf52/power_mgr_os/README.rst [hunk 1]
```diff
@@ -0,0 +1,92 @@
+.. _nrf52-power-mgr-sample:
```
Comment (dbkinder) [inline samples/boards/nrf52/power_mgr_os/README.rst]: This looks like it's a copy of an existing sample documentation with no obvious changes (including the title):  ``/samples/boards/nrf52/power_mgr/README.rst``.    The doc build complained that the label at the top of the doc here was the same as another document.

File: samples/boards/nrf52/power_mgr_os/README.rst [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+Requirements
+************
+
+This application uses nrf52 DK board for the demo.
+
+Building, Flashing and Running
+******************************
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/boards/nrf52/power_mgr
```
Comment (dbkinder) [inline samples/boards/nrf52/power_mgr_os/README.rst]: ... and the sample app you're running is in another folder, not the one with this documentation in it.

**Review by dbkinder — DISMISSED** [review]

File: arch/arm/soc/nordic_nrf/nrf52/power.c [hunk 3]
```diff
@@ -100,3 +100,23 @@ void _sys_soc_power_state_post_ops(enum power_states state)
...
+{
+	bool lps_flag;
+
+	switch (state) {
+	case SYS_POWER_STATE_CPU_LPS:
+	case SYS_POWER_STATE_CPU_LPS_1:
+#if defined(CONFIG_SYS_POWER_DEEP_SLEEP)
+	case SYS_POWER_STATE_DEEP_SLEEP:
+#endif
+		lps_flag = true;
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply arch/arm/soc/nordic_nrf/nrf52/power.c]: Yes will do.

File: arch/arm/soc/nordic_nrf/nrf52/power.c [hunk 4]
```diff
@@ -100,3 +100,23 @@ void _sys_soc_power_state_post_ops(enum power_states state)
...
+		break;
+	default:
+		lps_flag = false;
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply arch/arm/soc/nordic_nrf/nrf52/power.c]: yes, will do.

File: samples/boards/nrf52/power_mgr_os/README.rst [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+power management infrastructure to enter into Low Power state.
+
+This app will cycle through the following power schemes each time idle thread
+calls _sys_soc_suspend() hook function :
+
+1. Low Power State: Low Power State is SOC specific and being in this state is
+   transparent to devices. SOC and devices do not lose context in this Mode.
+   SOC supports the following two Low Power states :
+
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply samples/boards/nrf52/power_mgr_os/README.rst]: @pizi-nordic this is sample application documentation which i reused from existing sample. I will double check and fix the gaps.

Also on the PM states, from the documentation i saw that the nrf5x has support for two active power saving states (constant latency mode and low power mode).

The idea is to support pluggable PM policies based on the target application, in this base PR, I have added a simple policy to demonstrate the overall flows.

File: subsys/power/Kconfig [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	  - 1 ERROR, only write SYS_LOG_ERR
+	  - 2 WARNING, write SYS_LOG_WRN in addition to previous level
+	  - 3 INFO, write SYS_LOG_INF in addition to previous levels
+	  - 4 DEBUG, write SYS_LOG_DBG in addition to previous levels
+
+if SYS_POWER_LOW_POWER_STATE
+config PM_CONTROL_OS_LPS
+	bool "Platfrom supports LPS"
+	help
+	  This is the maximum number of bytes that the
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply subsys/power/Kconfig]: copy paste error, will fix it.

File: subsys/power/policy.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <power.h>
+#include <soc.h>
+#include <soc_power.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_PM_LEVEL
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply subsys/power/policy.c]: Ok will fix it.

File: subsys/power/device.c [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+#include <logging/sys_log.h>
+
+/*
+ * FIXME: Remove the conditional inclusion of
+ * core_devices array once we enble the capability
+ * to build the device list based on devices power
+ * and clock domain dependencies.
+ */
+#ifdef CONFIG_SOC_SERIES_NRF52X
+#define CONFIG_MAX_PM_DEVICES	15
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply subsys/power/device.c]: Ok will consider using the device list itself.

File: subsys/power/device.c [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+void sys_pm_resume_devices(void)
+{
+	int i;
+
+	for (i = 0; i < device_count; i++) {
+		if (!device_retval[i]) {
+			int idx = device_ordered_list[i];
+
+			device_set_power_state(&pm_device_list[idx],
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply subsys/power/device.c]: I think this kind of complexity can be handled in the driver itself, for example, on resume driver can submit worker to do the actual wake up and handle the device status accordingly.

File: subsys/power/device.c [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+		/* Check if the device is core device */
+		for (j = 0, is_core_dev = false; j < NUM_CORE_DEVICES; j++) {
+			if (!strcmp(pm_device_list[i].config->name, &core_devices[j][0])) {
+				is_core_dev = true;
+				break;
+			}
+		}
+
+		if (is_core_dev) {
```
[the remark this answers is not in the document]
Comment (ramakrishnapallala) [reply subsys/power/device.c]: As you pointed rightly, we need to use device tree here to handle the device dependencies which will come as incremental changes in the upcoming patches.

File: samples/boards/nrf52/power_mgr_os/README.rst (hunk 1 above)
Comment (ramakrishnapallala) [reply samples/boards/nrf52/power_mgr_os/README.rst]: yes David, will fix it.

File: samples/boards/nrf52/power_mgr_os/README.rst (hunk 2 above)
Comment (ramakrishnapallala) [reply samples/boards/nrf52/power_mgr_os/README.rst]: Ok will fix it.

**Comment by ramakrishnapallala** [comment]
@pizi-nordic  @dbkinder  I have addressed your comments and updated the PR, please have a look.

**Comment by ramakrishnapallala** [comment]
@dbkinder I am getting following warning:

/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/doc/_build/rst/doc/samples/subsys/power/power_mgr/README.rst: WARNING: document isn't included in any toctree

How to fix this?

**Comment by nashif** [comment]
> How to fix this?

```
diff --git a/samples/subsys/power/power.rst b/samples/subsys/power/power.rst
new file mode 100644
index 0000000000..e1dd2cc6ee
--- /dev/null
+++ b/samples/subsys/power/power.rst
@@ -0,0 +1,10 @@
+.. _power_management-samples:
+
+Power Management Samples
+########################
+
+.. toctree::
+   :maxdepth: 1
+   :glob:
+
+   **/*
```

File: subsys/power/pm_policy.h [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+extern "C" {
+#endif
+
+/**
+ * @brief Function to create device PM list
+ */
+extern  void sys_pm_create_device_list(void);
+
+/**
+ * @brief Function to supend the devices in PM device list
```
Comment (rgundi) [inline subsys/power/pm_policy.h]: nitpick - spelling mistake in suspend

File: subsys/power/policy.c [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE();
+
+#ifdef CONFIG_TICKLESS_KERNEL
+#define SECS_TO_TICKS		1000
+#else
+#define SECS_TO_TICKS		100
+#endif
+
```
Comment (rgundi) [inline subsys/power/policy.c]: Isn't the ticks configurable? So, shouldn't you use a macro function to address it here?

File: samples/boards/nrf52/power_mgr_os/README.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply samples/boards/nrf52/power_mgr_os/README.rst]: Are you referring the "Sub power modes" in http://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.nrf52832.ps.v1.1%2Fpower.html ?

If so, you see that mentioned power modes only affect PPI interconnect delay and CPU wakeup time. Changing these settings based on predicted sleep time, which is done in the example, is not a thing which we can recommend as these settings are in fact driven by the real-time requirements of application (we might break operation of PPI-based solutions relying on constant event latency if we switch to low power mode in runtime).

Despite of the fact, that this feature is a bit hard to show in nRF5x family, I think, that OS-level power management is a important feature which have to be implemented in Zephyr.

File: subsys/power/device.c [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+{
+
+	for (int i = device_count - 1; i >= 0; i--) {
+		int idx = device_ordered_list[i];
+
+		/* If necessary  the policy manager can check if a specific
+		 * device in the device list is busy as shown below :
+		 * if(device_busy_check(&device_list[idx])) {do something}
+		 */
+		device_retval[i] = device_set_power_state(&pm_device_list[idx],
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply subsys/power/device.c]: I think, that until blocking still is there, this feature will be not really usable in real-word applications.

File: subsys/power/device.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply subsys/power/device.c]: Ok. This will also make the sleep state longer saving more time. What about applications which need small latency?

**Comment by pizi-nordic** [comment]
>SYS_LOG: I looked at the ./doc/subsystems/logging/logger.rst but an example in kernel or driver code will be really helpful.

Please look at https://github.com/zephyrproject-rtos/zephyr/pull/8816. I think that this is the most comprehensive example.  There are also few PRs which are smaller (like https://github.com/zephyrproject-rtos/zephyr/pull/9361), but do not show all features.

**Comment by nashif** [comment]
@pizi-nordic are you ok with the changes otherwise? I think the PR was changed to use new logging subsystem already and we can further optimize the usage in the stabilisation period after this gets merged as part of 1.13

**Comment by pizi-nordic** [comment]
@pdunaj: FYI

**Comment by pizi-nordic** [comment]
@nashif: The biggest issue what I see is bonding of CPU/PPI latency mode to the predicted sleep time. IMHO we should completely remove code which alters this setting from the `nrf52/power.c`.

I assume that rest of the pointed things will be solved later in subsequent PRs, so I have no problem with them.

File: subsys/power/policy.c [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+#define LOG_MODULE_NAME power
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE();
+
+#ifdef CONFIG_TICKLESS_KERNEL
+#define SECS_TO_TICKS		CONFIG_TICKLESS_KERNEL_TIME_UNIT_IN_MICRO_SECS
+#else
+#define SECS_TO_TICKS		CONFIG_SYS_CLOCK_TICKS_PER_SEC
```
Comment (pizi-nordic) [inline subsys/power/policy.c]: Such tick <-> time conversion is imprecise on some SoCs (expecially nRF5x). If you need a more precise timing, please use `__ticks_to_ms()` and `_ms_to_ticks()`.

**Comment by ramakrishnapallala** [comment]
@pizi-nordic 
"The biggest issue what I see is bonding of CPU/PPI latency mode to the predicted sleep time. IMHO we should completely remove code which alters this setting from the nrf52/power.c."

_sys_soc_set_power_state from nrf52/power.c is only offering a mechanism to enter different power states(SYS_POWER_STATE_CPU_LPS/SYS_POWER_STATE_CPU_LPS_1) and does not have any info on power state residency or latency requirement. It is the PM policy which is making decisions on which PM state to enter through  _sys_soc_suspend which is being defined by the application in the current code. The proposed changes are built on top of them.

So the application designer or developer has to define a PM policy which suits his requirements, like if the system/CPU is going to idle for X secs(residency), then the system can enter into LPS. And these are configurable and can be changed based on application needs.

**Review by nashif — APPROVED** [review]

File: subsys/power/device.c [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+ */
+#ifdef CONFIG_SOC_SERIES_NRF52X
+#define MAX_PM_DEVICES	15
+#define NUM_CORE_DEVICES	4
+#define MAX_DEV_NAME_LEN	16
+static const char core_devices[NUM_CORE_DEVICES][MAX_DEV_NAME_LEN] = {
+	"clk_k32src",
+	"clk_m16src",
+	"sys_clock",
+	"UART_0",
```
Comment (pdunaj) [inline subsys/power/device.c]: UART may be disabled

**Pull request merged**